Add changelog for 1.2.0 release, update version in CMake (#629)
heremaps/here-olp-sdk-cpp: CHANGELOG.md

## v1.2.0 (04/02/2020)

**Common**

* Added a protected read-only cache. For details, see `CacheSettings`.

> * Added a protected read-only cache [usage example](https://github.com/heremaps/here-olp-sdk-cpp/blob/master/docs/dataservice-cache-example.md).

Reviewer: This will not work forever. Probably locking to a specific commit is better, e.g. https://github.com/heremaps/here-olp-sdk-cpp/blob/c30c4051ec012eaca09b7aa7ba4c7e8f6cd7a576/docs/dataservice-cache-example.md
Last commit before this change.

* Added ARM platform support for Embedded Linux with GCC 5.4 (32/64 bit).
* Added the `ErrorCodeToString` function to olp/core/http/NetworkTypes.h.

> * `NetworkCurl` logging is now improved.

Reviewer 2: ```suggestion
* `Improved the NetworkCurl` logging.
```

* Added the bool operator to the `HRN` class. You can now use the operator to check whether all the service type fields in the `HRN` class are not empty.
* Fixed a possible data race in the `Condition` class.
* Added the `MoveResult` method to the `ApiResponse` class. The method returns an rvalue reference to the result.
* CMake now requires RapidJSON v1.1.0.
* Various CMake cleanups.
* Improved error handling in Android `HttpClient`.
* Fixed the linking error when SDK was statically linked inside a shared library with `CMAKE_POSITION_INDEPENDENT_CODE=ON`.
* The retry policy now handles server errors automatically.

**olp-cpp-sdk-authentication**

* Added a possibility to parse credentials from a file or stream in `olp::authentication::AuthenticationCredentials`. Now, to get credentials for the `AuthenticationCredentials` class, you can use the **credentials.properties** file that is provided by the platform.
* Deprecated the `TokenEndpoint` class. It will be removed by 04.2020.
* Deprecated the `TokenRequest` class. It will be removed by 04.2020.
* Deprecated various methods in `AuthenticationClient`.

> * Authentication now gets the current timestamp from the platform and uses it for requesting the oauth2 token to prevent replay attacks. As a fallback, the system time is used.

Reviewer 2: ```suggestion
* Authentication now gets the current timestamp from the platform and uses it to request the OAuth2 token that prevents replay attacks. As a fallback, the system time is used.
```

**olp-cpp-sdk-dataservice-read**

> * The `QueryApi`, `MetadataApi`, `LookupApi` and `BlobApi` are now blocking, and use the new sync `CallApi` method in `OlpClient`.

Reviewer 2: ```suggestion
* The `QueryApi`, `MetadataApi`, `LookupApi`, and `BlobApi` classes are now blocking and use the new synchronous `CallApi` method in `OlpClient`.
```

* The `GetData` method of the versioned and volatile layer now triggers the `PreconditionFailed` error if both a partition ID and data handle are passed in the data request.

> * **Work-in-progress**: Add reading streamed data support via new `StreamLayerClient`. Currently only subscribing, unsubscribing and getting data from a stream layer message is implemented.

Reviewer 2: ```suggestion
* **Work-in-progress**: Added reading support for streamed data using the new `StreamLayerClient`. Currently, you can only subscribe, unsubscribe, and get data from a stream layer message.
```

## v1.2.0 (04/02/2020)

**Common**

* Add protected read-only cache, see `CacheSettings` for details.
* Add protected read-only cache usage example.
* Add a new platform support Embedded Linux (32/64 bit).
* Add `ErrorCodeToString` function to olp/core/http/NetworkTypes.h.
* Curl logging improved (unexpected messages are logged).
* IsNull function in `HRN` class was changed to handle realms according to platform changes.
* Add operator bool to `HRN` class, used to check if the `HRN` has all service type relevant fields not empty.
* Fix a possible data race in the `Condition` class.
* Add `MoveResult` to `ApiResponse`, it returns rvalue reference to result.
* CMake now requires RapidJSON v1.1.0.
* Various CMake cleanups.
* Android `HttpClient` error handling improved.

> * Fixed a linking error when SDK was statically linked inside shared library. (CMAKE_POSITION_INDEPENDENT_CODE=ON)

Reviewer 2: Should users specify CMAKE_POSITION_INDEPENDENT_CODE=ON when they build a project? If yes, please specify.

* Retry policy now handles server errors automatically.
* New retry back off algorithm introduced. now uses jitter algorithm.

**olp-cpp-sdk-authentication**

* Add a possibility to parse credentials from file or stream, in `olp::authentication::AuthenticationCredentials`. Now users can use the credentials.properties directly, which is provided by the platform.
* `TokenEndpoint` class is deprecated.
* `TokenRequest` class is deprecated.
* Various methods in `AuthenticationClient` are deprecated.
* Authentication now gets the current timestamp from the platform.

**olp-cpp-sdk-dataservice-read**

> * `QueryApi`, `MetadataApi`, `LookupApi`, `BlobApi`, is now blocking, and use a new sync `CallApi` method in `OlpClient`.

Reviewer: ```suggestion
* `QueryApi`, `MetadataApi`, `LookupApi` and `BlobApi` are now blocking, and use the new sync `CallApi` method in `OlpClient`.
```

> * `QueryApi`, `MetadataApi`, `LookupApi`, `BlobApi`, is now blocking, and use a new sync `CallApi` method in `OlpClient`.

Reviewer 2: Did I get it right?

> * Versioned layer and volatile layer `GetData` method will trigger an PreconditionFailed error if both partition id and data handle is passed in the data request.

Reviewer: ```suggestion
* Versioned layer and volatile layer `GetData` method will trigger a PreconditionFailed error if both partition id and data handle are passed in the data request.
```

> * Add reading streamed data support via new `StreamLayerClient`.

Reviewer: ```suggestion
* **Work-in-progress**: Add reading streamed data support via new `StreamLayerClient`. Currently only subscribing, unsubscribing and getting data from a stream layer message is implemented.
```

**olp-cpp-sdk-dataservice-write**

> * Stream Layer client enhanced to use task scheduler for async operations instead of network thread.

Reviewer: ```suggestion
* Stream Layer client enhanced to use `TaskScheduler` for async operations instead of network thread.
```
